sdk/eventhub/azure-eventhub/azure/eventhub/aio/_buffered_producer/_buffered_producer_dispatcher_async.py

```diff
@@ -0,0 +1,141 @@
+
+    async def enqueue_events(self, events, *, partition_id=None, partition_key=None, timeout=None):
+        pid = await self._get_partition_id(partition_id, partition_key)
+        async with self._lock:
+            try:
+                await self._buffered_producers[pid].put_events(events, timeout)
+            except KeyError:
+                buffered_producer = BufferedProducer(
+                    self._create_producer(partition_id=pid),
+                    pid,
+                    self._on_success,
+                    self._on_error,
+                    self._max_message_size_on_link,
+                    max_wait_time=self._max_wait_time,
+                    max_concurrent_sends=self._max_concurrent_sends,
+                    max_buffer_length=self._max_buffer_length
+                )
+                await buffered_producer.start()
+                self._buffered_producers[pid] = buffered_producer
+                await buffered_producer.put_events(events, timeout)
+
+    async def flush(self, timeout=None):
+        # flush all the buffered producer, the method will block until finishes or times out
+        async with self._lock:
+            futures = []
+            for pid, producer in self._buffered_producers.items():
+                # call each producer's flush method
+                futures.append((pid, asyncio.ensure_future(producer.flush(timeout=timeout))))
+
+            # gather results
+            exc_results = {}
+            for pid, future in futures:
+                try:
+                    await future
+                except Exception as exc:  # pylint: disable=broad-except
+                    exc_results[pid] = exc
+
+            if not exc_results:
+                _LOGGER.info("Flushing all partitions succeeded")
+                return
+
+            _LOGGER.warning('Flushing all partitions partially failed with result %r.', exc_results)
+            raise EventDataSendError(
+                message="Flushing all partitions partially failed, failed partitions are {!r}"
+                        " Exception details are {!r}".format(exc_results.keys(), exc_results)
+            )
+
+    async def close(self, *, flush=True, timeout=None, raise_error=False):
+
+        futures = []
+        # stop all buffered producers
+        for pid, producer in self._buffered_producers.items():
+            futures.append((pid, asyncio.ensure_future(
+                producer.stop(flush=flush, timeout=timeout, raise_error=raise_error)
+            )))
+
+        exc_results = {}
+        # gather results
+        for pid, future in futures:
+            try:
+                await future
+            except Exception as exc:  # pylint: disable=broad-except
+                exc_results[pid] = exc
+
+        if exc_results:
+            _LOGGER.warning('Stopping all partitions failed with result %r.', exc_results)
+            if raise_error:
+                raise EventDataSendError(
```

Comment:
Should this be an EventDataSendError?

Reply:
I changed this to `EventHubError`.

I've logged a separate issue to track the error experience.

Reply:
https://github.com/Azure/azure-sdk-for-python/issues/24011